agent/runner/actions/mongodb_explain_action.go

```diff
@@ -89,28 +98,152 @@ func (a *mongodbExplainAction) Run(ctx context.Context) ([]byte, error) {
 	}
 	defer client.Disconnect(ctx) //nolint:errcheck
 
-	var eq proto.ExampleQuery
+	return explainForQuery(ctx, client, a.params.Query)
+}
+
+func (a *mongodbExplainAction) sealed() {}
+
+func (e explain) prepareCommand() (bson.D, error) {
+	command := e.Command
+	switch e.Op {
+	case "query":
+		if len(command) == 0 {
+			command = e.Query
+		}
+
+		if len(command) == 0 || command[0].Key != "find" {
+			var filter any
+			if len(command) != 0 && command[0].Key == "query" {
+				filter = command[0].Value
+			} else {
+				filter = command
+			}
+
+			return bson.D{
+				{Key: "find", Value: e.getCollection()},
+				{Key: "filter", Value: filter},
+			}, nil
+		}
+
+		return dropDBField(command), nil
+	case "update":
+		if len(command) == 0 {
+			command = bson.D{
+				{Key: "q", Value: e.Query},
+				{Key: "u", Value: e.UpdateObj},
+			}
+		}
+
+		return bson.D{
+			{Key: "update", Value: e.getCollection()},
+			{Key: "updates", Value: []any{command}},
+		}, nil
+	case "remove":
+		if len(command) == 0 {
+			command = bson.D{{Key: "q", Value: e.Query}}
+		}
+
+		return bson.D{
+			{Key: "delete", Value: e.getCollection()},
+			{Key: "deletes", Value: []any{command}},
+		}, nil
+	case "getmore":
+		if len(e.OriginatingCommand) == 0 {
+			return bson.D{{Key: "getmore", Value: ""}}, nil
+		}
+
+		command = e.OriginatingCommand
+
+		return dropDBField(command), nil
+	case "command":
+		command = dropDBField(command)
+
+		if len(command) == 0 || command[0].Key != "group" {
+			return command, nil
+		}
+
+		return fixReduceField(command), nil
+	// not supported
+	case "insert", "drop":
+		return nil, errors.Errorf("command %s is not supported for explain", e.Op)
+	}
+
+	return command, nil
+}
+
+func (e explain) getDB() string {
+	s := strings.SplitN(e.Ns, ".", 2)
+	if len(s) == 2 {
+		return s[0]
+	}
+
+	return ""
+}
+
+func (e explain) getCollection() string {
+	s := strings.SplitN(e.Ns, ".", 2)
+	if len(s) == 2 {
+		return s[1]
+	}
+
+	return ""
+}
 
-	err = bson.UnmarshalExtJSON([]byte(a.params.Query), true, &eq)
+func dropDBField(command bson.D) bson.D {
+	for i := range command {
+		if command[i].Key != "$db" {
+			continue
+		}
+
+		if len(command)-1 == i {
+			return command[:i]
+		}
+
+		return append(command[:i], command[i+1:]...)
+	}
+
+	return command
+}
+
+func fixReduceField(command bson.D) bson.D {
```

Comment:
Could you please add comments to each function why we are doing that?

Reply:
RN and comments added.